stl/inc/xutility

```diff
@@ -4823,24 +4852,130 @@ public:
+
+    // using multiply-shift instead of modulo to fairly map values in range [0, _Mod), adapted from arXiv:1805.10941v4
+    static constexpr _Udiff _Reduce(_Udiff _Value, _Udiff _Mod, size_t _Bitrange) noexcept {
+        constexpr auto _Udiff_bits = CHAR_BIT * sizeof(_Udiff);
+        constexpr auto _Uint_bits  = CHAR_BIT * sizeof(uint32_t);
+        // _Bitrange may not exceed the number of bits in result type
+        _STL_INTERNAL_CHECK(_Bitrange <= _Udiff_bits);
+        if (_Bitrange > _Uint_bits) {
+            uint64_t _Val_hi = 0;
+            uint64_t _Val_lo = __rng_umul128(_Value, _Mod, &_Val_hi);
+
+            if (_Bitrange < _Udiff_bits) {
+                _Val_hi = __rng_shiftright128(_Val_lo, _Val_hi, static_cast<uint32_t>(_Bitrange));
             }
+            return static_cast<_Udiff>(_Val_hi);
+        } else if (_Bitrange > _Uint_bits / 2) {
+            // ensure promotion to ull to keep all bits
+            return static_cast<_Udiff>((static_cast<uint64_t>(_Value) * _Mod) >> _Bitrange);
+        }
+
+        // avoid unnecessary promotion to 64-bit for smaller types
+        return static_cast<_Udiff>((static_cast<uint32_t>(_Value) * _Mod) >> _Bitrange);
+    }
+
+    _NODISCARD static constexpr bool _Is_power_of2(_Udiff _Value) noexcept {
+        return _Value && !(_Value & (_Value - 1));
+    }
 
-            // _Ret is [0, _Mask], _Index - 1 <= _Mask, return if unbiased
-            if (_Ret / _Index < _Mask / _Index || _Mask % _Index == _Udiff(_Index - 1)) {
-                return static_cast<_Diff>(_Ret % _Index);
+    _NODISCARD _Diff operator()(const _Diff _Index) { // adapt _Urng closed range to [0, _Index)
+        constexpr auto _Udiff_bits = CHAR_BIT * sizeof(_Udiff);
+        const auto _UIndex         = static_cast<_Udiff>(_Index);
+        _Udiff _Ret                = _Get_bits(); // random bits
+        _Udiff _Mask               = _Bmask; // 2^N - 1, _Ret is within [0, _Mask]
+        size_t _Bitrange           = _Bits;
+
+        while (_Mask < static_cast<_Udiff>(_Index - 1)) { // need more random bits
+            // if loop is entered, _Bits can't shift out of range
+            _Ret <<= _Bits;
+            _Ret |= _Get_bits();
+            _Mask <<= _Bits;
+            _Mask |= _Bmask;
+            _Bitrange += _Bits;
+        }
+
+        // exceeding bits from _Get_bits() that don't fit into _Ret are discarded,
+        // therefore _Bitrange can never be larger than _Udiff_bits
+        if (_Bitrange > _Udiff_bits) {
+            _Bitrange = _Udiff_bits;
+        }
+
+        _Udiff _Threshold = _Mask - (_Mask % _UIndex);
+        for (;;) { // try a sample random value
+            // _Ret is [0, _Mask], _UIndex - 1 <= _Mask, return if unbiased
+            if (_Ret < _Threshold || _Is_power_of2(_UIndex)) {
+                return static_cast<_Diff>(_Reduce(_Ret, _UIndex, _Bitrange));
             }
+            // re-use some bits of the previous iteration

```

Comment:
Bits of the previous iteration are no longer uniformly distributed given the condition that they are rejected. _Bitrange new bits need to be generated for each iteration.

Reply:
Will be changed.